### PR TITLE
Allow ALPAKA_CXX_STANDARD to propagate to nvcc with MSVC 1920 and above

### LINKED ISSUE #1129: ALPAKA_CXX_STANDARD is not passed to nvcc on Windows
I have C++17 code using alpaka that compiles fine on Ubuntu with nvcc 11. However, the same code fails to compile on Windows, with nvcc 11 emitting errors on C++17 features.

`ALPAKA_CXX_STANDARD` is set to 17, but it seems this information is not passed forward to nvcc. If I add `-std=c++17` manually to the `CUDA_NVCC_FLAGS`, the code compiles fine on Windows.

### PR DESCRIPTION
**Pull request by bernhardmgruber**
Fixes #1129

### CODE REVIEW

**Review by sbastrakov — DISMISSED** [review]
Looks good to me.
However, I am not sure why MSVS was guarded against in the first place. @BenjaminW3 do you have objections?

**Comment by BenjaminW3** [comment]
Older Visual Studio versions did not support those flags. This may have been the reason.

**Comment by bernhardmgruber** [comment]
> Older Visual Studio versions did not support those flags. This may have been the reason.

That is surprising to me since I would expect `CUDA_NVCC_FLAGS` to go directly to `nvcc` and not affect VS.

So for VS2019 it does not seem to be an issue. The CI for VS2017 lists this warning:
`CUSTOMBUILD : nvcc warning : The -std=c++14 flag is not supported with the configured host compiler. Flag will be ignored.`
So it just seems to cause a warning with older versions of VS. This is acceptable for me, since not having the flag prevents C++17 from being available in CUDA (nvcc defaults to C++14 with MSVC, if the flag is not set), which is far worse.

**Comment by BenjaminW3** [comment]
Can you guard it so that the warning does not appear?

**Comment by BenjaminW3** [comment]
Could you please add `ALPAKA_CXX_STANDARD: 17` randomly to 2 `nvcc + MSVC` build jobs in `.github/workflows/ci.yml`?
This is the new feature that you are adding and we should test it.

**Comment by BenjaminW3** [comment]
With the current change we are getting the warning `CUSTOMBUILD : nvcc warning : The -std=c++14 flag is not supported with the configured host compiler. Flag will be ignored.` for every compilation unit. I do not think that this is acceptable.

**Comment by bernhardmgruber** [comment]
> Could you please add `ALPAKA_CXX_STANDARD: 17` randomly to 2 `nvcc + MSVC` build jobs in `.github/workflows/ci.yml`?
> This is the new feature that you are adding and we should test it.

It seems we first need CI builds with CUDA 11 and MSVC. Before that, C++17 is not supported in CUDA. This is addressed here: #1086

**Comment by bernhardmgruber** [comment]
By now I start wondering why we even need `ALPAKA_CXX_STANDARD` different from `CMAKE_CXX_STANDARD`? Why are we not just passing `CMAKE_CXX_STANDARD` to nvcc?

File: cmake/alpakaCommon.cmake [hunk 1]
```diff
@@ -460,9 +460,7 @@ if(ALPAKA_ACC_GPU_CUDA_ENABLE)
                     )
                 endforeach()
 
-                if(NOT MSVC)
-                    list(APPEND CUDA_NVCC_FLAGS -std=c++${ALPAKA_CXX_STANDARD})
-                endif()
+                list(APPEND CUDA_NVCC_FLAGS -std=c++${ALPAKA_CXX_STANDARD})
```
Comment (psychocoderHPC) [inline cmake/alpakaCommon.cmake]: You can cover this line with `MSVC_VERSION` to keep the old behaviour.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -150,55 +150,55 @@ jobs:
...
+          env: {CXX: cl.exe,  CC: cl.exe, ALPAKA_CI_CL_VER: 2019,                                   CMAKE_BUILD_TYPE: Debug,   ALPAKA_CI_BOOST_BRANCH: boost-1.73.0, ALPAKA_CI_CMAKE_VER: 3.18.0,                     ALPAKA_CXX_STANDARD: 17,  ALPAKA_ACC_GPU_CUDA_ENABLE: ON, ALPAKA_CUDA_VERSION: "11.0", ALPAKA_CUDA_ARCH: "35;75", ALPAKA_ACC_GPU_CUDA_ONLY_MODE: ON,                                                                ALPAKA_ACC_CPU_B_SEQ_T_FIBERS_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_THREADS_ENABLE: OFF, ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLE: OFF, ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE: OFF, ALPAKA_ACC_CPU_BT_OMP4_ENABLE: OFF}
```
Comment (bernhardmgruber) [inline .github/workflows/ci.yml]: here I added `ALPAKA_CXX_STANDARD: 17`

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -150,55 +150,55 @@ jobs:
...
           os: windows-2019
-          env: {CXX: cl.exe,  CC: cl.exe, ALPAKA_CI_CL_VER: 2019,                                   CMAKE_BUILD_TYPE: Release, ALPAKA_CI_BOOST_BRANCH: boost-1.68.0, ALPAKA_CI_CMAKE_VER: 3.18.0,                     ALPAKA_ACC_GPU_CUDA_ENABLE: ON, ALPAKA_CUDA_VERSION: "11.0",                                                                                                                              ALPAKA_ACC_CPU_B_SEQ_T_FIBERS_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE: OFF, ALPAKA_ACC_CPU_BT_OMP4_ENABLE: OFF, ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLE: OFF}
+          env: {CXX: cl.exe,  CC: cl.exe, ALPAKA_CI_CL_VER: 2019,                                   CMAKE_BUILD_TYPE: Release, ALPAKA_CI_BOOST_BRANCH: boost-1.68.0, ALPAKA_CI_CMAKE_VER: 3.18.0,                     ALPAKA_CXX_STANDARD: 17,  ALPAKA_ACC_GPU_CUDA_ENABLE: ON, ALPAKA_CUDA_VERSION: "11.0",                                                                                                                              ALPAKA_ACC_CPU_B_SEQ_T_FIBERS_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE: OFF, ALPAKA_ACC_CPU_BT_OMP4_ENABLE: OFF, ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLE: OFF}
```
Comment (bernhardmgruber) [inline .github/workflows/ci.yml:204]: here I added `ALPAKA_CXX_STANDARD: 17`

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -150,55 +150,55 @@ jobs:
...
+          env: {CXX: cl.exe,  CC: cl.exe, ALPAKA_CI_CL_VER: 2017,                                   CMAKE_BUILD_TYPE: Release, ALPAKA_CI_BOOST_BRANCH: boost-1.67.0, ALPAKA_CI_CMAKE_VER: 3.17.3,                                               ALPAKA_ACC_GPU_CUDA_ENABLE: ON, ALPAKA_CUDA_VERSION: "10.0", ALPAKA_CUDA_ARCH: "30;75", ALPAKA_ACC_GPU_CUDA_ONLY_MODE: ON, ALPAKA_CUDA_NVCC_SEPARABLE_COMPILATION: ON,                    ALPAKA_ACC_CPU_B_SEQ_T_FIBERS_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_THREADS_ENABLE: OFF, ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLE: OFF, ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE: OFF, ALPAKA_ACC_CPU_BT_OMP4_ENABLE: OFF}
```
Comment (psychocoderHPC) [inline .github/workflows/ci.yml]: What do you changed here? I do mot see a difference but it ia marked as different.

Comment (bernhardmgruber) [reply .github/workflows/ci.yml]: I changed the white-space so the columns in the entire file are properly aligned again.

**Comment by psychocoderHPC** [comment]
We run into a internal compiler error: https://github.com/alpaka-group/alpaka/pull/1130/checks?check_run_id=1085102857#step:4:2958

```
 -- Build files have been written to: D:/a/alpaka/alpaka/build
     Creating library D:/a/alpaka/alpaka/build/test/integ/matMul/Release/matMul.lib and object D:/a/alpaka/alpaka/build/test/integ/matMul/Release/matMul.exp
  matMul.vcxproj -> D:\a\alpaka\alpaka\build\test\integ\matMul\Release\matMul.exe
  Building NVCC (Device) object test/unit/math/CMakeFiles/math.dir/src/Release/math_generated_math.cpp.obj
  math.cpp
CUSTOMBUILD : nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning). [D:\a\alpaka\alpaka\build\test\unit\math\math.vcxproj]
Error #3246 : Internal Compiler error (codegen): "there was an error in verifying the lgenfe output!" [D:\a\alpaka\alpaka\build\test\unit\math\math.vcxproj]
  
CUSTOMBUILD : nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning). [D:\a\alpaka\alpaka\build\test\unit\math\math.vcxproj]
  math.cpp
  CMake Error at math_generated_math.cpp.obj.Release.cmake:280 (message):
    Error generating file
    D:/a/alpaka/alpaka/build/test/unit/math/CMakeFiles/math.dir/src/Release/math_generated_math.cpp.obj
```

**Comment by bernhardmgruber** [comment]
I have seen it last week, but I cannot reproduce it. All tests compile and run fine on my Windows machine with VS 2019, nvcc 11 and ALPAKA_CXX_STANDARD 17.

**Comment by psychocoderHPC** [comment]
We need to reproduce the issue else we will break the CI :-(

**Comment by psychocoderHPC** [comment]
Maybe we should try to use the latest CUDA 11 version http://developer.download.nvidia.com/compute/cuda/11.0.3/local_installers/cuda_11.0.3_451.82_win10.exe in https://github.com/alpaka-group/alpaka/blob/39580a1b2d438fa5df1d65ee9a40c1dadf850d92/script/install_cuda.sh#L69
@bernhardmgruber could you update this line and push again.

**Comment by bernhardmgruber** [comment]
That is a good hint, thank you! I upgraded to CUDA 11.0.3 and rebased onto develop.

**Comment by psychocoderHPC** [comment]
error still exists :-(

**Comment by bernhardmgruber** [comment]
I could finally reproduce the issue, but I have no solution to it:
```
cmake .. -DALPAKA_CXX_STANDARD=17 -DALPAKA_ACC_GPU_CUDA_ENABLE=ON -DBUILD_TESTING=ON
cmake --build . --target math

Microsoft (R) Build Engine version 16.7.0+b89cb5fde for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/dev/alpaka/test/catch_main/CMakeLists.txt
  CatchMain.cpp
  CatchMain.vcxproj -> C:\dev\alpaka\build_win\test\catch_main\Debug\CatchMain.lib
  Building Custom Rule C:/dev/alpaka/test/common/CMakeLists.txt
  Dummy.cpp
  common.vcxproj -> C:\dev\alpaka\build_win\test\common\Debug\common.lib
  Building NVCC (Device) object test/unit/math/CMakeFiles/math.dir/src/Debug/math_generated_math.cpp.obj
  math.cpp
CUSTOMBUILD : nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning). [C:\dev\alpaka\build_win\test\unit\math\math.vcxproj]
Error #3246 : Internal Compiler error (codegen): "there was an error in verifying the lgenfe output!" [C:\dev\alpaka\build_win\test\unit\math\math.vcxproj]

CUSTOMBUILD : nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning). [C:\dev\alpaka\build_win\test\unit\math\math.vcxproj]
  math.cpp
  CMake Error at math_generated_math.cpp.obj.Debug.cmake:280 (message):
    Error generating file
    C:/dev/alpaka/build_win/test/unit/math/CMakeFiles/math.dir/src/Debug/math_generated_math.cpp.obj
```

**Comment by psychocoderHPC** [comment]
CUDA 11.1 is now released but the issue is still existing :-(  (I already tried it in my own fork)

**Comment by bernhardmgruber** [comment]
Thanks a lot for trying!

**Comment by bernhardmgruber** [comment]
I also added CUDA 11.1 to the CI for Windows.

**Comment by bernhardmgruber** [comment]
I took the time to hunt the problem down and the offending code causing the issue is passing `alpaka::test::unit::math::Buffer` to the `TestKernel::operator()` in `alpaka/test/unit/math/src/math.cpp`. Thinking about this, this seems really weird.

`alpaka::test::unit::math::Buffer` contains the following data members:
* `alpaka::DevCpu`
* `alpaka::Buf<alpaka::DevCpu, ...>`
* `alpaka::Buf<DevAcc, ...>`
* `TData * const pHostBuffer`
* `TData * const pDevBuffer`

All this is passed to the GPU kernel. Should we even support this? Passing a host device and a host buffer and a pointer to a host buffer into the GPU kernel? This is eventually triggering the codegen error when we enable C++17 on MSVC.

**Comment by sbastrakov** [comment]
I don't think alpaka buffers have a device interface at all, so passing an object that has a member of type `alpaka::Buf` to a kernel should not work in my opinion.

**Comment by bernhardmgruber** [comment]
With #1193 merged, this should work now ;)

**Review by sbastrakov — APPROVED** [review]

**Pull request merged**